### PR TITLE
Kill jsr-305

### PR DESCRIPTION
**Pull request by jtnord**
remove JSR-305 annotations from core.

See [JENKINS-55973](https://issues.jenkins-ci.org/browse/JENKINS-55973).

### Proposed changelog entries

* Developer: Switched bug detection annotations from JSR-305 to SpotBugs / net.jcp equivalents

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [x] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [x] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Comment by jtnord** [comment]
Not sure how we can prevent someone adding an annotation back in (esp when the jsr-305 annotations are used in dependencies (and we still would like spotbugs to detect issues with the use of those libraries...)

**Comment by timja** [comment]
> Not sure how we can prevent someone adding an annotation back in (esp when the jsr-305 annotations are used in dependencies (and we still would like spotbugs to detect issues with the use of those libraries...)

checkstyle can do it https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/imports/IllegalImportCheck.html

**Comment by timja** [comment]
Are you aware of: https://github.com/google/guava/issues/2960#issuecomment-595946238 ?

**Comment by jtnord** [comment]
> checkstyle can do it https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/imports/IllegalImportCheck.html

I was tying to avoid checkstyle - as that sounds from the outset like it would be more contentious to introduce - even with a single rule.. But I'll give it whirl and see what the feedback is

> Are you aware of: [google/guava#2960 (comment)](https://github.com/google/guava/issues/2960#issuecomment-595946238) ?

I am now :-)  I think there is value in this as-is  even if here could be another migration when that project comes to light and the tooling is updated to support it.  (because we would be migrating from two sets of annotations not three)

**Comment by jtnord** [comment]
@timja added checkstyle to prevent re-introduction.

File: pom.xml
Comment (timja) [inline pom.xml]: would be nice to have this in a file, 
here's an example of how we got it working in multi module with jcasc:
https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/pom.xml#L82

non-blocking though

**Review by timja — APPROVED** [review]
I've done a cursory look through, but the diff is huge, one comment on checkstyle but non-blocking

File: pom.xml
Comment (jtnord) [reply pom.xml]: Thanks for the pointer. 
Given I have no idea if tools (eclipse intellij netbeans etc) all cope with that correctly (given it is set by the mvn wrapper) - and not everyone imports all the projects in a multi module setup, I'll stick which this for now as I know it to work.

**Comment by jglick** [comment]
> Not sure how we can prevent someone adding an annotation back in

Add to `bom` a `dependencyManagement` entry pinning that artifact at a visibly bogus `version`. Which is probably how we should be handling `javax.servlet:servlet-api`, rather than using the `[0]` trick.

File: war/pom.xml
Comment (jglick) [inline war/pom.xml]: Why? Please revert.

**Review by jglick — COMMENTED** [review]
For the benefit of NetBeans users, please

```diff
diff --git core/src/main/resources/META-INF/upgrade/JSR-305.hint core/src/main/resources/META-INF/upgrade/JSR-305.hint
new file mode 100644
index 0000000000..14c9a4227c
--- /dev/null
+++ core/src/main/resources/META-INF/upgrade/JSR-305.hint
@@ -0,0 +1,3 @@
+javax.annotation.CheckForNull => edu.umd.cs.findbugs.annotations.CheckForNull;;
+javax.annotation.Nonnull => edu.umd.cs.findbugs.annotations.NonNull;;
+javax.annotation.Nullable => edu.umd.cs.findbugs.annotations.Nullable;;
```

**Review by jeffret-b — APPROVED** [review]
I'm in favor of this cleanup. I also did a cursory review of some of the files. Everything I saw looked good. I didn't review everything though.

**Comment by jtnord** [comment]
> > Not sure how we can prevent someone adding an annotation back in
> 
> Add to `bom` a `dependencyManagement` entry pinning that artifact at a visibly bogus `version`. Which is probably how we should be handling `javax.servlet:servlet-api`, rather than using the `[0]` trick.

handled by checkstyle at the import level.  want to keep them available for third party libs that use them.

File: war/pom.xml
Comment (jtnord) [reply war/pom.xml]: lord knows..  I edited this file via eclipse dependency panel and not by hand.

**Review by Wadeck — APPROVED** [review]
Not looked into details but I am really pleased to see such inconsistency corrected. It will ease the future PRs to choose their annotation.

@stephenc  will be happy I imagine :)

**Comment by fcojfernandez** [comment]
@jtnord have a look at the CI:

```
/home/jenkins/workspace/Core_jenkins_PR-4604/core/src/main/java/jenkins/management/ShutdownLink.java:[71,5]
```

In master there are new annotated methods so this PR cannot be merged:
https://github.com/jenkinsci/jenkins/blob/8c1e567e7d400e2e51ab8f9bc7cf90ca8b0290c8/core/src/main/java/jenkins/management/ShutdownLink.java#L65-L75

**Comment by jtnord** [comment]
people need to stop merging stuff then 😂

**Comment by jtnord** [comment]
> For the benefit of NetBeans users, please
> 
> ```diff
> diff --git core/src/main/resources/META-INF/upgrade/JSR-305.hint core/src/main/resources/META-INF/upgrade/JSR-305.hint
> new file mode 100644
> index 0000000000..14c9a4227c
> --- /dev/null
> +++ core/src/main/resources/META-INF/upgrade/JSR-305.hint
> @@ -0,0 +1,3 @@
> +javax.annotation.CheckForNull => edu.umd.cs.findbugs.annotations.CheckForNull;;
> +javax.annotation.Nonnull => edu.umd.cs.findbugs.annotations.NonNull;;
> +javax.annotation.Nullable => edu.umd.cs.findbugs.annotations.Nullable;;
> ```

Addressed in af115fd

**Review by jglick — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]

**Comment by daniel-beck** [comment]
Care needs to be taken when merging this due to a potential conflict with other PRs merged around the same time.

I would recommend merging other PRs towards the weekly, and then, once that is released, merge master into this to get confirmation everything's OK, and merge this afterwards very early next week.

**Review by oleg-nenashev — APPROVED** [review]
It has to happen, sooner or later. It is highly likely that other pull requests will be affected by merge conflicts, but all of them should be trivial ones. I suggest merging it after the weekly is out today. Will merge in 24 hours if nobody is against

**Comment by timja** [comment]
> It has to happen, sooner or later. It is highly likely that other pull requests will be affected by merge conflicts, but all of them should be trivial ones. I suggest merging it after the weekly is out today. Will merge in 24 hours if nobody is against

I was planning on merging this as soon as the weekly was out, and not merging anything else till this in 😄

File: core/src/main/java/hudson/tasks/ArtifactArchiver.java
Comment (jtnord) [inline core/src/main/java/hudson/tasks/ArtifactArchiver.java:116]: then surely this should be a `boolean` :-o  (not a comment on the PR this a WTF comment)

Comment (timja) [reply core/src/main/java/hudson/tasks/ArtifactArchiver.java:116]: 🤦‍♂

**Comment by jtnord** [comment]
> jtnord deleted a comment from timja 1 minute ago

@timja sorry deleted my comment which seems to have deleted yours.  the `Boolean` was for a read resolve - which checked it for null - so it is a Boolean but obviously can not be `NonNull` as `readResolve` for old versions would not work - and SpotBugs should be warning about a redundant null check... 🤷‍♂

File: war/pom.xml
Comment (timja) [inline war/pom.xml:84]: FYI this broke JCasC, I've created a PR to fix it: https://github.com/jenkinsci/configuration-as-code-plugin/pull/1359

**Pull request merged**
